operator/cluster/cluster.go

```diff
@@ -202,6 +190,56 @@ func AddClusterBase(clientset *kubernetes.Clientset, client *rest.RESTClient, cl
 
 }
 
+//getOrCreatePVC will first verify whether the pgcluster name matches the "current primary" CRD value of the pgcluster CRD.
+//Based on this information, the underlying PVC name for the primary pod can be validated. If the expected PVC exists,
+//the correct PVC name is returned.
+//If the "current primary" label and the pgcluster name do match, this is either the recreation of a cluster that has not
+//previously failed over, or it is a new pgcluster. If a recreation, the PVC is reused, otherwise a new PVC is allocated.
+func getOrCreatePVC(clientset *kubernetes.Clientset, cl *crv1.Pgcluster, namespace string) string {
+	//if the current primary label does not equal the base pgcluster name and it is not empty,
+	//then the pgcluster has failed over at some point
+	if cl.Annotations[config.ANNOTATION_CURRENT_PRIMARY] != cl.Spec.Name && cl.Annotations[config.ANNOTATION_CURRENT_PRIMARY] != "" {
+		//ensure the expected PVC exists
+		_, found, err := kubeapi.GetPVC(clientset, cl.Annotations[config.ANNOTATION_CURRENT_PRIMARY], namespace)
+		if err != nil {
+			log.Errorf("error while trying to find the PVC %s during creation of cluster %s. Error: %s", cl.Annotations[config.ANNOTATION_CURRENT_PRIMARY], cl.Spec.Name, err)
+			publishClusterCreateFailure(cl, err.Error())
+			return ""
+		}
+		if found {
+			log.Debugf("pvc [%s] already present from previous cluster with this same name, will not recreate", cl.Annotations[config.ANNOTATION_CURRENT_PRIMARY])
+			return cl.Annotations[config.ANNOTATION_CURRENT_PRIMARY]
+		}
+	}
+
+	//otherwise, if the current primary name matches the PVC name, this cluster is either new, or has not previously had a failover
+	//also, since the PVC not being found is returned as an error and the relevant errors are logged in GetPVC, error
+	//information will not be captured here
+	_, found, err := kubeapi.GetPVC(clientset, cl.Spec.Name, namespace)
+	//if the error is not null, and it is not a not found error, we will want to log the error and stop cluster creation
+	//otherwise, we can continue to either reuse or create the PVC with the 'standard' name
+	if err != nil && !kerrors.IsNotFound(err) {
+		log.Errorf("error while trying to find the PVC %s during creation of cluster %s. Error: %s", cl.Spec.Name, cl.Spec.Name, err)
+		publishClusterCreateFailure(cl, err.Error())
+		return ""
+	}
+	//if the default PVC name is found, then we will not recreate it. The existing PVC will be used for this pgcluster
+	if found {
+		log.Debugf("pvc [%s] already present from previous cluster with this same name, will not recreate", cl.Spec.Name)
+		return cl.Spec.Name
+	} else {
```

Comment:
We don't need an `else` here. the `if found` block returns

Reply:
This function has been rewritten